SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityScheduledJobPoller.java

```diff
@@ -28,110 +37,161 @@
+            return request.getRequest().getScheduledExpectedRuntimeMillis();
+        } else {
+            final Optional<SingularityDeployStatistics> deployStatistics = deployManager.getDeployStatistics(taskId.getRequestId(), taskId.getDeployId());
+
+            if (deployStatistics.isPresent() && deployStatistics.get().getAverageRuntimeMillis().isPresent()) {
+                return deployStatistics.get().getAverageRuntimeMillis();
+            }
+            long nextRunAtTimestamp = 0;
+            final long now = System.currentTimeMillis();
+            try {
+                if (request.getRequest().getScheduleTypeSafe() == ScheduleType.RFC5545) {
+                    final String schedule = request.getRequest().getSchedule().get();
+                    final RecurrenceRule recurrenceRule = new RecurrenceRule(schedule);
+                    if (recurrenceRule.isInfinite()) {
+                        // set limit at 2100-01-01 00:00:00
+                        recurrenceRule.setUntil(new DateTime(2100, 1, 1, 0, 0, 0));
+                    }
+
+                    // DTSTART is RFC5545 but NOT in the recur string, but its a nice to have
+                    Pattern pattern = Pattern.compile("DTSTART=([0-9]{8}T[0-9]{6})");
+                    Matcher matcher = pattern.matcher(schedule);
+                    DateTime startDateTime;
+                    org.joda.time.DateTime dtStart;
+                    if (matcher.find()) {
+                        DateTimeFormatter formatter = DateTimeFormat.forPattern("yyyyMMdd'T'HHmmss");
+                        dtStart = formatter.parseDateTime(matcher.group(1));
+                    } else {
+                        dtStart = org.joda.time.DateTime.now();
+                        dtStart = dtStart.withSecondOfMinute(0);
+                    }
+
+                    startDateTime = new DateTime(dtStart.getYear(), (dtStart.getMonthOfYear() - 1), dtStart.getDayOfMonth(),
+                        dtStart.getHourOfDay(), dtStart.getMinuteOfHour(), dtStart.getSecondOfMinute());
+
+                    RecurrenceRuleIterator timeIterator = recurrenceRule.iterator(startDateTime);
+                    nextRunAtTimestamp = 0;
+                    while (timeIterator.hasNext()) {
+                        nextRunAtTimestamp = timeIterator.nextMillis();
+                        if (nextRunAtTimestamp >= now) {
+                            break;
+                        } else {
+                            nextRunAtTimestamp = 0;
+                        }
+                    }
+                    if (nextRunAtTimestamp <= 0) {
+                        String msg = String.format("No next run date found for %s (%s)", taskId, request.getRequest().getScheduleTypeSafe());
+                        LOG.warn(msg);
+                        exceptionNotifier.notify(msg, ImmutableMap.of("taskId", taskId.toString()));
+                        return Optional.absent();
+                    }
```

Comment:
I actually thought about doing that too. I'll take care of it and commit

I'm not a Java guy, can you tell me where I should place the new RFC5545Schedule.java file into the folder structure? Should I put it inside com.hubspot.singularity folder inside the SingularityService module?